auth/oauth/types.go

```diff
@@ -20,20 +20,83 @@
 package oauth
 
 import (
+	"encoding/json"
 	"github.com/nuts-foundation/nuts-node/core"
 	"net/url"
 )
 
 // this file contains constants, variables and helper functions for OAuth related code
 
-// TokenResponse is the OAuth access token response
+// TokenResponse is the OAuth access token response.
+// Through Set() and GetString() additional parameters (for OpenID4VCI, for instance) can be set and retrieved.
 type TokenResponse struct {
 	AccessToken string  `json:"access_token"`
 	ExpiresIn   *int    `json:"expires_in,omitempty"`
 	TokenType   string  `json:"token_type"`
-	CNonce      *string `json:"c_nonce,omitempty"`
 	Scope       *string `json:"scope,omitempty"`
-	Status      *string `json:"status,omitempty"`
+
+	additionalParams map[string]interface{}
+}
+
+var _ json.Unmarshaler = (*TokenResponse)(nil)
+var _ json.Marshaler = (*TokenResponse)(nil)
+
+func (t *TokenResponse) UnmarshalJSON(data []byte) error {
+	type Alias TokenResponse
+	var result Alias
+	// base parameters
+	if err := json.Unmarshal(data, &result); err != nil {
+		return err
+	}
+	// extension parameters
+	additionalParams := map[string]interface{}{}
+	_ = json.Unmarshal(data, &additionalParams) // can't fail, already unmarshalled
+	delete(additionalParams, "access_token")
+	delete(additionalParams, "expires_in")
+	delete(additionalParams, "token_type")
+	delete(additionalParams, "scope")
+	*t = TokenResponse(result)
+	if len(additionalParams) > 0 {
+		t.additionalParams = additionalParams
+	}
+	return nil
+}
+
+func (t TokenResponse) MarshalJSON() ([]byte, error) {
+	result := make(map[string]interface{})
+	for key, value := range t.additionalParams {
+		result[key] = value
+	}
+	result["access_token"] = t.AccessToken
+	result["expires_in"] = t.ExpiresIn
+	result["token_type"] = t.TokenType
+	result["scope"] = t.Scope
+
+	return json.Marshal(result)
+}
+
+// Set adds an additional parameter to the token response.
+// It's a builder-style function.
+func (t *TokenResponse) WithParam(key string, value interface{}) *TokenResponse {
```

Comment:
is using this to set keys `access_token`, `expires_in`, `token_type`, or `scope` incorrect usage (godoc) or should it set the struct param? (same for the getter)

Reply:
Illegal usage. I considered allowing it, but that leads to a lot of extra code and basically changes it to a map with helper functions that validate. It's how the JWX library works (with headers, for instance), but although it's elegant, usage is awkward because of the errors that can be returned.

So `With` and `Get` are only for the parameters not defined in the OAuth2 RFC. I added that as godoc.